### ISSUE TITLE
The wrapper conundrum

### ISSUE THREAD

**Issue by Yomguithereal** [medialab/graphlib#8]
Ok, this one was actually brought by @jacomyal @jacomyma @boogheta and myself:

Should we wrap the data given by the user. Meaning that, when you get a node, do you get precisely what you gave, or a wrapper (class instance, or overloaded object), or not.

```js
// All examples below will instantiate the graph thusly
var graph = new Graph({
  nodes: [{id: 1, label: 'One'}, {id: 2, label: 'Two'}],
  edges: [{source: 1, target: 2}] 
});

// Example n°1: no wrapping
var node = graph.get(1);
// returns the original node, unscathed
>>> {
  id: 1,
  label: 'One'
}
// if you want to get neighbours
graph.neighbours(1);
// or
graph.neighbours(node);

// Example n°2: overloading
var node = graph.get(1);
>>> {
  id: 1,
  label: 'One',
  ...SomeMoreThings
}
// if you want to get neighbours
node.neighbours();

// Example n°3: class instance or wrapping
var node = graph.get(1);
>>> node instanceof Node
node.data
>>> {
  id: 1,
  label: 'One'
}
// if you want to get neighbours
node.neighbours();
```

Every solution (apart from n°2, which we should really avoid I think) has its merit.

* N°1 is the purest but makes the API a little more verbose
* N°3 might imply more RAM usage but has a more fluent API and has a serializarion cost

If you have other solutions, be free to chime in :)

**Comment by jacomyal** [comment]
I like the solution 1 better, but without the `graph.method(nodeReference)`. It looks more "functional" and less "OOP" than the 3rd one to me - more idiomatic of JavaScript, somehow.

IMO sigma's graph model accessors were actually quite satisfying:
```javascript
graph.nodes(); // returns a copy of the full nodes array
graph.nodes(nodeId); // returns node
graph.nodes(nodeId1, nodeId2); // returns { nodeId1: node1, nodeId2: node2 }
graph.nodes([nodeId1, nodeId2]); // returns [node1, node2]
// And the same goes for edges as well
```

**Comment by jacomyma** [comment]
n°2 should still be discussed. Though it's not my favorite, it's not clear to me that its drawbacks/benefits ratio pushes it out of consideration.

I propose to list "trials" or test cases and discuss if thay are important or not, and which solutions pass it or not.

#### A. Retrieving my initial data
```javascript
// jsonGraph similar to previous examples
var richGraph = new Graph(jsonGraph);
...doStuff // add nodes, modify attributes...

richGraph.getGraph();
// Returns the same JSON nodes and edges
>>> {nodes:[...], edges:[...]}

richGraph.getNodes();
// Returns the same JSON nodes
>>> [...]
```
Does it work? n°1 Yes / n°2 Nope / n°3 Sort of
Is it really important? I'm not convinced. I probably need to see examples where it causes an issue.

#### B. Asking data to items themselves
```javascript
// Get neighbors of neighbors
var nOfNIndex = {};
// Solution n°1
graph.getNeighbors(node).forEach(n1 => {
  graph.getNeighbors(n1).forEach(n2 => {
    nOfN[n2.id] = n2;
  });
});
// Note: I think .getNeighbors() makes more sense than .neighbors() in this context

// Solution n°2 & n°3
node.neighbors().forEach(n1 => {
  n1.neighbors().forEach(n2 => {
    nOfN[n2.id] = n2;
  });
});
```
Does it work? n°1 Ugly / n°2 OK / n°3 OK
For me n°1 is more difficult to read.

#### C. Consistency of accessing items' attributes
In this example we want to compute the degree of nodes at different moments. For instance we will remove some nodes and we want to keep a trace of the original degree of nodes. The goal of this example is to compare how we access an attribute compared to how we access a method.
```javascript
// Solution n°1
var degreeOld = node.degree1;
var degreeNow = graph.degree(node);

// Solution n°2
var degreeOld = node.degree1;
var degreeNow = node.degree();

// Solution n°3
var degreeOld = node.data.degree1;
var degreeNow = node.degree();
```
Does it work? n°1 Badly / n°2 OK / n°3 Acceptable
I like the solution n°2 better here. Solution n°3 is acceptable to me though. I really dislike solution n°1 despite using the simplest way of writing it: ```graph.degree(node)``` could easily end up ```graph.getNodeDegree(node)```.

**Comment by jacomyal** [comment]
I believe there are some cases where having different sources of information (the `Graph`, `Node` and `Edge` classes and related methods from solution 2 and 3) can be a problem, that are easier to solve the solution 1:

```javascript
var graph = new Graph({ nodes: [{id: 1, label: 'One'}], edges: [] });
var node = graph.getNode(1);
graph.dropNode(1);

// The fact that I still can manipulate a reference to "node" looks very weird to me:
console.log(node);
    // Solution 2: {id: 1, label: 'One'} + prorotype
    // Solution 3: Some object

console.log(node.getDegree()); // ?
```

With solution 1, now:
```javascript
var graph = new Graph({ nodes: [{id: 1, label: 'One'}], edges: [] });

var node = graph.getNode(1);
graph.dropNode(1); // or graph.dropNode(node), whatever

console.log(node);
    // {id: 1, label: 'One'}, ie the data as it was in the graph when retrieved

console.log(graph.getDegree(node.id)); // Returns undefined, or throws
```

These cases are easier to solve with the solution 1: Devs shouldn't store and manipulate the data provided by `Graph` methods too long, but just retrieve them when needed, to ensure the data they are reading are the ones as they are in the graph, when the methods are called.

If instead I work with instances of objects, I will want to actually keep them and manipulate them later, as I do with the DOM for instance - like when people cache jQuery selectors results during their apps lifecycle to avoid having to search the results again.

**Comment by jacomyma** [comment]
This is a point. I think JS is not Java though, and I feel like our position should be "use-it-wisely" rather than "idiot-dev-proof". It does not mean that ambiguities are fine, but I think that it's more important to propose good practices than to prevent bad uses. ```graph.getDegree()``` makes it clear that the degree is computed live, but at the cost of making simple things more complicated.

I think that for in simple cases a ```Graph``` should behave like an array. The array has the issue you talk about:

```javascript
var arr = new Array({id: 1, label: 'One'});

var node = arr[0];
delete arr[0];

console.log(node);
>>> {id: 1, label: 'One'} // ie the data as it was in the array when retrieved

console.log(arr[0]);
>>> undefined
```

**Comment by jacomyal** [comment]
> our position should be "use-it-wisely" rather than "idiot-dev-proof"
> it's more important to propose good practices than to prevent bad uses

We assisted together last Sunday at FOSDEM a [conference](https://fosdem.org/2016/schedule/event/design_linux_kernel_api/) about design errors in the Linux kernel-user-space API. The speaker [expressed](https://fosdem.org/2016/schedule/event/design_linux_kernel_api/attachments/slides/1204/export/events/attachments/design_linux_kernel_api/slides/1204/designing_linux_kernel_APIs_fosdem_2016_Kerrisk.pdf) two interesting statements:

  - User-space programmers are endlessly inventive
  - Try to imagine the ways in which an army of inventive user-space programmers might (ab)use your API

##### Latest illustration of that I have in mind:

Some months ago, I updated [Baobab](https://github.com/Yomguithereal/baobab) to the version 2 in most of my projects, and my whole codebases broke, because *I had been misusing Baobab's API from the beginning*, and nothing made me realize before this update.

I was coding like this:

```javascript
//   1. Extract some data:
var data = tree.get('data');

//   2. Mutate it in some way:
data.docs = data.docs.filter(isDocIn);

//   3. Set the fresh new data back into the tree:
tree.set('data', data);
```

Since Baobab v2+, objects extracted from the tree are frozen by default, and here, the `data.docs` assignement started throwing errors.

I was using Baobab the way I wanted to, and I did not even realize it was not the actually expected way to use it.

My point being: I believe **"idiots" will always find a way**, and I don't think "use-it-wisely" is enough.

---

> but at the cost of making simple things more complicated

I see the things quite differently: The solution 1 provides **one unique source** of data, when the other solutions actually provide three, with three different APIs. So, the code you would write might be a bit less simple, but at the cost of keeping the API design simpler, with less concepts and less ambiguities.

**Comment by jacomyma** [comment]
I understand your concern but I personally draw the opposite conclusion. Users will always find a way to fuck up, even with additional protections. I believe that the best strategy is to have as much consistency as possible, and simple examples to disseminate good practices. I don't see your baobab example as an API issue but as a lack of consistency: what you did made sense to you because that's what you do in other contexts.

**Comment by Yomguithereal** [comment]
After discussing the issue with @paulgirard, here are some other points:

* We should stick with the performant implementation and n°1 seems to be more so than n°3
* The location where we should store neighbourhood indexes is important. If we store it within the Graph object, then implementation should mimick this fact and n°1 it is. If we store it in at node level, then n°3 seems more natural.
* Having the Node instances rely on an internal reference to the Graph itself or store some indexes themselves could produce memory leaks.

**Comment by Yomguithereal** [comment]
I guess we are also reaching some questions raised by #2 here. If we represent this library as being on the same level as the JavaScript native Array or Set, for instance, they we shouldn't try to prevent the user from fucking up. If you fuck up your Array, well, you fucked up. Issue #11 is another problem altogether.

**Comment by Yomguithereal** [comment]
@DavidBruant, what's your opinion on this point?

**Comment by jacomyma** [comment]
I'd like us to consider the feasibility of being consistent JS native Array and so. I consider this kind of consistency as a major benefit for this kind of lib, which we expect simple at the core. However not at any cost, notably performance. Other kinds of consistency are also involved (internal consistency).

Also, n°1 being more performant seems a bold claim to me (I just don't see why).

**Comment by Yomguithereal** [comment]
If you want to get consistent with JS native array, then you shouldn't wrap items. When you give something to an array it does not give you back an `Item` instance with special things such as `item.next()` or `item.prev()` :grinning:.

**Comment by jacomyma** [comment]
Maybe it's not clear that I'm talking about ```.map()```, ```.forEach()``` and so on.

**Comment by jacomyma** [comment]
Ok what about something different:
```javascript
// Nodes and edges are just arrays
var n = graph.nodes[0];

// Nodes and edges are just objects
n;
>>> {id:0, label:'foo'}

// But we can wrap everything
$(n).degree; // or .degree()
>>> 0

$(n).neighbors;
>>> []

// A possible way to deal with ambiguities
$(n).label;
>>> undefined

$(n).attr('label');
>>> 'foo'

// $ is just for the sake of this example
```

**Comment by jacomyal** [comment]
If this `$` wrapper is global and not related to a particular graph, then it will cause problems when using multiple graph instances. 

Also, it would probably be necessary to have different wrappers for nodes and edges.

**Comment by Yomguithereal** [comment]
I think @jacomyma means that `$` is actually your graph instance. So no global wrapper issues.

```js
var graph = new Graph(...);

graph(node).degree;
```

**Comment by jacomyma** [comment]
I was thinking like @Yomguithereal but @jacomyal's point is real. Both options make sense in terms of writing. I don't see all the implications though. Does the ```graph(node).degree``` option totally solve the point?

I would enjoy this kind of writing, leveraging ```g``` as a short name for both the graph and a wrapper:
```javascript
var g = new Graph(...);
var node = g.nodes[0];

node.label;
>>> foo

g(node).degree();
>>> 0
```

**Comment by Yomguithereal** [comment]
One problem though is that I cannot see a way for a JS class instance to be used as a function also. What I mean is that:

```js
var g = new Graph(...);
g(node).degree();
```

Just cannot work (except if you fiddle dark things and maybe extends the Function prototype which we should really not do).

This could

```js
var graph = new Graph(...);
var g = graph.wrapper;
g(node).degree();
```

**Comment by Yomguithereal** [comment]
More about the wrapper thing:

1. if you consider the wrapper to be global and not instance specific then you have an issue because your node need to be both wrapped in a specific class + have a reference to its parent graph which could produce memory leaks.
2. if you consider the wrapper to be instance specific, then, it can be derived from solution n°1 very easily with some partial applications and bindings.

**Comment by jacomyma** [comment]
I would favor this compromise. The n°1+wrapper strategy has a unique set of benefits:
- Nodes and edges are normal arrays, which enforces the idea that the graph structure is simple, and then using it is very consistent. We don't have the ```node.data``` oddity.
- We don't have the clumsy ```graph.getDegree(node)``` and it feels like we are manipulating nodes thanks to ```g(node).degree()```. At the cost of often writing with this wrapper, which is somehow similar to jQuery, we have the feeling of working with nodes and edges themselves and I feel it's very consistent as well. I appreciate the similarity and nonetheless clear difference between ```node.label``` and ```g(node).degree()```.
- The only oddity I see for now is the required line to get the wrapper: ```var g = graph.wrapper``` We can probably find something shorter but it is not a big deal anyway since you do not have to write that a lot. I imagine that if it is in all our examples, users will just reproduce the pattern.

**Comment by Yomguithereal** [comment]
One issue I found with solution n°3 while trying it on some examples this afternoon is the following:

Let's say you create a graph, you create bunch of nodes & edges etc. and you want to get the neighbours of one of your nodes, you will need an intermediate reference to do so and this feels weird.

```js
var graph = new Graph();

// I create my node
var node = {
  id: 'whatever',
  label: 'A random node'
};

graph.addNode(node);

// ... we add some edges

// Now let's iterate over its neighbours
// First we need to get the internal node instance to do so
// even though we already have a reference on it above
var internalNode = graph.getNode('whatever');

internalNode.neighbours().forEach(...);

// This is quirky. You have two different references to what should be a same thing
// You have both a node variable and another one which is 'internalNode' or 'realNode'
// or whatever you need
```

With solution n°1 it feels more natural:

```js
var graph = new Graph();

// I create my node
var node = {
  id: 'whatever',
  label: 'A random node'
};

graph.addNode(node);

// ... we add some edges

// Now let's iterate over its neighbours
graph.neighbours(node).forEach(...);
```

**Comment by Yomguithereal** [comment]
> I would favor this compromise. The n°1+wrapper strategy has a unique set of benefits:

I agree. What's more both n°1 and wrapper can coexist.

> We don't have the clumsy graph.getDegree(node)

I don't find it clumsy. It is simple and has compositional advantages. Let's say you want to map the degree of every nodes in the graph:

```js
// You can do the following
var degrees = graph.nodes.map(graph.getDegree);

// With n°3 you need to do:
var degrees = graph.nodes.map(function(node) {
   return node.degree();
});

// With wrapper
var degress = graph.nodes.map(function(node) {
  return g(node).degree();
});
```

**Comment by jacomyma** [comment]
First I have to precise that I'm essentially discussing lib usability issues, which I know are not the only concern. I know that solution n°1 is the clean one but I still think it's necessary to debate about the writing style it delivers, because it's so important for a lib.

About solution n°3's drawback: I think your argument is not a good one, I'm not trolling and I think it's important to share that point because it will come over and over. For any lib you can have (and will have) bad practices. The existence of bad practices or ambiguities cannot be avoided, and arguably should not. What is really important is that the recommended practice (guidelines and examples) are clear and easily understandable. Preventing bad uses is done by disseminating good uses, not by imposing constraints. I believe constraints have another purpose (that point is off-topic). Your example is of this kind: solution n°3 allows a bad use, but that bad use could be easily discouraged by a simpler good practice:
```javascript
// BAD
var node = {
  id: 'whatever',
  label: 'A random node'
};
graph.addNode(node);

// GOOD
graph.addNode( {
  id: 'whatever',
  label: 'A random node'
});
// Explanation: keeping the reference to the original object creates issues
```

**TLDR; the existence of a bad practice is not really an issue. However the inexistence of a good practice is the real deal. We should focus on good examples rather than bad ones.**

In the same perspective ```graph.getDegree()``` is marginally useful but I would not advise constraining users to do it instead of something like ```g(node).degree()``` if we can. We may like the first one for some reason but I think its clumsiness is nonetheless real:
```javascript
// Not clumsy
thing.property; // in JS you read "thing" as the important object
yomgui.age;

// Clumsy
somethingHigher.getPropertyOf(thing); // emphasizes "somethingHigher"
metricsProvider.getAge(yomgui);

// Close to not clumsy
$(thing).property; // emphasizes "$(thing)"
$(yomgui).age;
```
We should not hide the writing style issue just because we have reasons to favor other solutions. Not confusing style and implementation is key to successfully discussing the lib.

**Comment by Yomguithereal** [comment]
```js
// BAD
var node = {
  id: 'whatever',
  label: 'A random node'
};
graph.addNode(node);
```

Why is that a bad practice? What if you need to dynamically create the node? What if some of its attributes should be set conditionally? What if the node preexist the graph because you got the data from somewhere else? What if you indexed this data elsewhere? After all, a node is just a data point and can exist outside of the graph and live in multiple locations simultaneously.

Therefore, the degree of the node is only relevant within a precise graph context. For instance, and I am not trolling either, if we insert an object in a list, its index in the list is known by the list alone and has only a relevance in this list's context. The object doesn't know about its index in the list and you have to do `array.indexOf(object)` to retrieve this piece of information.

**Comment by Yomguithereal** [comment]
On the topic of wrappers now, there are still some issues:

* You need two different wrappers

```js
var graph = new Graph(...);
var gn = graph.nodeWrapper;
var ge = graph.edgeWrapper;

gn(node).neighbours();
ge(edge).target(); // or whatever

// You cannot have a single wrapper since it means
// you would need to rely on strange heuristics
// needing to guess which kind of object was passed to you
// g(o) => is o a node? is o an edge?
// you can find a efficient way with two hashmaps (less efficient than raw object)
// and you would still need to poll the two to find the solutions
```

* Wrappers are less performant:

```js
var graph = Graph();

// with n°1
graph.getDegree(node);
>>> 1 func call

// with n°2
var g = graph.wrapper
g(node).degree();
// or
g(node).degree;
>>> 1 or 2 func calls, 1 object allocated (probably with additional stuff you don't need right now)

* Wrappers are complex. Wrappers means more abstractions to handle whereas you could only have a single one, the `Graph`.

```js
// Do you really find the following:
var graph = new Graph(...);
graph.getDegree(node);

// More complex than:
var graph = new Graph(...);
var $ = graph.wrapper;
$(node).degree;
```

* It is hard to use composition and functional paradigms with wrappers.

* No JS native item works this way. Why can't we imagine a `Graph` abstraction as simple as a native `Array` where you give, handle and get back raw JavaScript data?

**Comment by jacomyma** [comment]
@Yomguithereal about the example with a "bad practice". I just mean that *we can make it* a bad practice. We can discourage something if we want, mostly by favoring a "good practice" instead and explaining the issue. It does not mean that it is bad in general or in other contexts. Publishing a lib requires to communicate a lot about how it's intended to be used. d3 leveraged this a lot for instance.

**Comment by jacomyma** [comment]
The benefit of a wrapper does not appear in a short example. Here is a longer example that I find more explicit.
```javascript
// Without a wrapper
var graph = new Graph(...);
var node = graph.nodes[0];
graph.getDegree(node);
graph.getNeighbors(node).map(n => graph.getDegree(n));
graph.getOutEdges(node).map(e => {
  return graph.getDegree(graph.getTarget(e));
});

// With a wrapper
// ...supposing it returns wrapped objects
// ...supposing it is the same for nodes and edges
var graph = new Graph(...);
var $ = graph.wrapper;
var node = graph.nodes[0];
$(node).degree;
$(node).neighbors.map(n => n.degree);
$(node).outEdges.map(e => e.target.degree);
```
If I'm really the only one to prefer wrapping over graph getters, then so be it. By the way @Yomguithereal's issues are real. I hope solutions exist.

**Comment by Yomguithereal** [comment]
I believe we can mitigate the verbosity of your first example with a leaner API (not saying we should because it can make the API less "explicit"):

```js
// Without a wrapper
var graph = new Graph(...);
var node = graph.nodes[0];
graph.degree(node);
graph.neighbours(node).map(graph.degree);
graph.outEdges(node).map(e => graph.degree(e.target));
```

This said, it's not difficult to create a wrapper over solution 1's API in fact. My point is that it probably shouldn't be the lowest-level way of manipulating the graph.

> I'm admittedly addicted to chaining like node.outEdges[0].target.degree. The wrapper is just a way to allow it. I believe wrapperless chaining is also possible.

The concern is real. The only way I can find to replicate your example with n°1 is the following:

```js
graph.degree(graph.outEdges(node)[0].target)
```

You can somewhat chain but not completely.

**Comment by DavidBruant** [comment]
> Try to imagine the ways in which an army of inventive user-space programmers might (ab)use your API

I strongly recommend watching this https://www.youtube.com/watch?v=eL5o4PFuxTY on this topic (I have 5 other 40-60mins videos like this if you need more convincing ;-) ) This has nothing to do with security specifically. Let's be lazy and more importantly, let's force users to be lazy through a well-designed API.

> Users will always find a way to fuck up, even with additional protections.

Some protections are unbreachable. functions, API, these things work if defined properly.

````js
const f = (() => {
    const a = [1, 6, Math.random()]
    
    return x => Object.getOwnPropertyDescriptor(a, x).value
})();
````
It is not possible for anyone to modify the underlying array. No user will find a way ever. The ECMAScript language semantics guarantees it (let's forget about dumbnesses like the fact that Object.getOwnPropertyDescriptor is configurable by default).


> @DavidBruant, what's your opinion on this point?

I prefer the first version. No wrapping. The second idea (graph adds its information onto the nodes) seems dangerous and prevents using the same object to represent a node in two different graphs.
I know a [use case](https://github.com/DavidBruant/OoI/) where I need to represent a graph over time and be able to compare two versions across two different times. Being able to keep the same node is important for memory reasons (it's a big graph). In that context, the degree should be stored at the graph level because it refers to the node in relation to a specific graph.
Nodes can be immutable anyway.



> If this $ wrapper is global and not related to a particular graph, then it will cause problems when using multiple graph instances.
>
> Also, it would probably be necessary to have different wrappers for nodes and edges.

Agreed.


> ````js
> var g = new Graph(...);
> g(node).degree();
> ````
>
> Just cannot work (except if you fiddle dark things and maybe extends the Function prototype which > we should really not do).

It can work with something like:
````js
function Graph(){
    const nodes = new Set();
    const edges = new Set();
    
    const methods = {
        addNode(node){
            // ...
        }
        // ....
    }

    return Object.assign((node) => {
        // ....
    }), methods)
}
````
Not saying we should do it, but it can be done reasonably.
Given the data in nodes/edges will always be largely larger in memory than the method references (5, 10 objects ?), we can assume the loss by not sharing methods via prototype is not a big deal at all.

**Comment by boogheta** [comment]
my 2 cents, with no real argument, just an opinion
As a chaining user as well I like the idea of the wrapper, but I guess it makes sense to get it only as a complementary possibility. It would be some kind of sugar with a taste that only some like. So I guess going with option 1° with extra wrappers possible for anyone wanting it sounds best to me
